### PR TITLE
PT-1280: New gene and variant section

### PR DESCRIPTION
**Pull request by buske**
This PR updates the genotype section of the patient form by:
- Merging candidate, rejected, and solved genes within the same table
- Adding support for manually-entered candidate variants (PT-2023)
- Storing evidence and other information for candidate genes (PT-1381, PT-1149)

I made a few adjustments on top of @veronikaslc's base commit:
- a few bug fixes (mostly found after merging with remote-matching and patient-network functionality)
- removed formatting/style changes if spotted

Replaces #1873.

### CODE REVIEW

**Comment by buske** [comment]
@meatcar and @allasm, please review.

File: components/export/ui/src/main/resources/PhenoTips/ExportPreferences.xml [hunk 1]
```diff
@@ -131,11 +131,21 @@
        'creationDate' : "$services.localization.render('phenotips.exportPreferences.field.creationDate')",
        'author' : "$services.localization.render('phenotips.exportPreferences.field.author')",
        'date' : "$services.localization.render('phenotips.exportPreferences.field.date')",
-       'allergies' : "$services.localization.render('PhenoTips.AllergiesDataClass_allergies')",
```
Comment (buske) [inline components/export/ui/src/main/resources/PhenoTips/ExportPreferences.xml]: Should not have been removed?

File: components/patient-data/impl/src/main/resources/ApplicationResources.properties [hunk 2]
```diff
@@ -12,11 +12,7 @@ PhenoTips.PatientClass_family_history_hint=List health conditions found in famil
...
-PhenoTips.InvestigationClass_hint=List genes considered as candidates for the patient's disease.
-PhenoTips.RejectedGenesClass_hint=List genes that have been investigated and ruled out as candidates for the patient's disease.
 PhenoTips.PatientClass_solvedCase_hint=Check this box if the molecular diagnosis for the patient is known.
-PhenoTips.PatientClass_multipleGestation=Multiple gestation
```
Comment (buske) [inline components/patient-data/impl/src/main/resources/ApplicationResources.properties]: Should not be deleted.

File: components/patient-data/impl/src/main/resources/ApplicationResources.properties [hunk 3]
```diff
@@ -57,7 +53,7 @@ PhenoTips.PatientClass_global_mode_of_inheritance=Global mode of inheritance
 PhenoTips.PatientClass_indication_for_referral=Indication for referral
 PhenoTips.PatientClass_health_card=Provincial Health Card #
 PhenoTips.PatientClass_proband=This is the proband
-PhenoTips.PatientClass_family_of=This patient is part of the familiy of
```
Comment (buske) [inline components/patient-data/impl/src/main/resources/ApplicationResources.properties]: This change is unrelated and should be committed straight to master.

File: components/patient-data/indexing/src/main/java/org/phenotips/data/indexing/internal/PatientEmptyObjectsRemover.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+                if (field == null || StringUtils.isEmpty(field.getValue())) {
+                    doc.removeXObject(object);
+                    changed = true;
+                }
+            }
+        }
+
+        try {
+            if (changed) {
+                context.getWiki().saveDocument(doc, "Removed empty object", true, context);
```
Comment (buske) [inline components/patient-data/indexing/src/main/java/org/phenotips/data/indexing/internal/PatientEmptyObjectsRemover.java]: @sdumitriu Is this the correct way of handling a PatientChangingEvent where data is modified?

File: components/solr-access-service/ui/.gitignore [hunk 5]
```diff
@@ -0,0 +1 @@
+
```
Comment (buske) [inline components/solr-access-service/ui/.gitignore]: Remove.

File: components/widgets/api/src/main/resources/ApplicationResources.properties [hunk 6]
```diff
@@ -13,12 +13,13 @@ phenotips.imageDisplayer.done=Done
...
 
 phenotips.tableMacros.delete=Delete
-phenotips.tableMacros.reject=Reject
 phenotips.tableMacros.newEntry=New entry
+phenotips.tableMacros.newVariantEntry=Add variant
 phenotips.tableMacros.noObjects=None specified
 phenotips.tableMacros.rowDeleteConfirmation=Are you sure you want to delete this row?
 phenotips.tableMacros.listNotFound=Cannot find the list to update
-phenotips.tableMacros.rejectConfirmation=Are you sure you want to reject this row?
+phenotips.tableMacros.typeNotFound=Cannot add data: type not found
```
Comment (buske) [inline components/widgets/api/src/main/resources/ApplicationResources.properties]: Unclear for user (and me).

File: components/widgets/ui/src/main/resources/PhenoTips/TabelarDataMacros.xml [hunk 7]
```diff
@@ -98,11 +94,10 @@ $xwiki.ssx.use('PhenoTips.TabelarDataMacros')##
...
-    {{html wiki="true" clean="false"}}#foreach($prop in $dataClass.properties)&lt;td class="${prop.name}#if ($prop.name == $options.propertyName) hidden#end"&gt;#if ("$!{options.labels}" == 'true'){{html clean="false" wiki="false"}}&lt;label &gt;${prop.translatedPrettyName}:&lt;/label&gt;{{/html}}#end$doc.display($prop.getName(), $options.mode, $targetObj)&lt;/td&gt;#end#if ($options.mode == 'edit')#if("$!{options.rejectToClass}" != '')&lt;td class="actions"&gt;#__extradata_rejectTool($targetObj '' $options)&lt;/td&gt;#end&lt;td class="actions"&gt;#__extradata_deleteTool($targetObj '')&lt;/td&gt;#end{{/html}}
+    {{html wiki="true" clean="false"}}#foreach($prop in $dataClass.properties)&lt;td class="${prop.name}#if ($prop.name == $options.propertyName) hidden#end"&gt;#if ("$!{options.labels}" == 'true'){{html clean="false" wiki="false"}}&lt;label&gt;${prop.translatedPrettyName}:&lt;/label&gt;{{/html}}#end$doc.display($prop.getName(), $options.mode, $targetObj)&lt;/td&gt;#end#if ($options.mode == 'edit')&lt;td class="actions"&gt;#__extradata_deleteTool($targetObj '')&lt;/td&gt;#end{{/html}}
   #end
 #end
-{{/velocity}}
-</content>
+{{/velocity}}</content>
```
Comment (buske) [inline components/widgets/ui/src/main/resources/PhenoTips/TabelarDataMacros.xml]: Irrelevant change.

File: components/widgets/ui/src/main/resources/PhenoTips/TabelarDataMacros.xml [hunk 8]
```diff
@@ -507,7 +464,6 @@ table.withLabel td, table.withLabel th {
 table.withLabel tr:first-child th {
   display: none;
 }
-
```
Comment (buske) [inline components/widgets/ui/src/main/resources/PhenoTips/TabelarDataMacros.xml]: Irrelevant change.

**Comment by buske** [comment]
I propose migrating CSS and JS to external resource files, as in https://github.com/phenotips/phenotips/blob/master/components/pedigree/ui/src/main/resources/PhenoTips/PedigreeEditor.xml

File: components/patient-data/indexing/src/main/java/org/phenotips/data/indexing/internal/PatientEmptyObjectsRemover.java (hunk 4 above)
Comment (sdumitriu) [inline components/patient-data/indexing/src/main/java/org/phenotips/data/indexing/internal/PatientEmptyObjectsRemover.java]: No, `*ingEvent` means that the document (as received in the `source` argument) can be modified and the code sending the event will save at the end.

File: components/widgets/api/src/main/resources/ApplicationResources.properties (hunk 6 above)
Comment (veronikaslc) [inline components/widgets/api/src/main/resources/ApplicationResources.properties]: That's exactly what we have in [TabelarDataMacros](https://github.com/phenotips/phenotips/blob/master/components/widgets/ui/src/main/resources/PhenoTips/TabelarDataMacros.xml) line 238. I'll change to "Cannot add data: variant object not found"

**Comment by veronikaslc** [comment]
For migration CSS and JS, Gene/Variants section doesn't seem to be a standalone widget like Pedigree and thers no so much of a massive CSS with it. It's implemented as alternative way of constructing table to TabelarDataMacros, more-so, it's implemented on top of it in terms of CSS. So moving CSS and JS files for it, if decided, should logically happen for both simultaneously, as I see it, thus, should go as a separate commit.

**Comment by buske** [comment]
Oops, I've already moved them out and will be pushing a rebased version of this commit shortly. If people want to put those resources back into the xmls, or separate more out, we can do that as well.

**Comment by buske** [comment]
Oh no, @veronikaslc, it looks like we're doing double work! I'm going through and implementing these fixes, too, and will commit them all shortly. Sorry for not being clear about that...

**Comment by buske** [comment]
I addressed all the comments, rebased, and split the PR up into five sections:
ce6d1d2ec95f5593955d3e0e021ff1f34531ff6e - UI and core functionality
37b805287fb1213e4701e63eb27cd8db88997393 - Listeners to remove empty objects
0591e26858307bd4de6a1191a40c79b63f4b3eb0 - Controllers
b8efd66bda89b00b3407afc1b63edd71033b1732 - Migrations
234b9a46a2cba27af62036b2525e1d2bb5387ca9 - Excel export

Patient index tests were originally included in this PR, and have now been split off into PR #1915.

**Comment by veronikaslc** [comment]
No worries, I just implemented and commited changes you 've mentioned in the pull request for PT 1280  thread lately.

File: components/widgets/resources/src/main/resources/resources/uicomponents/widgets/extraGeneVariantData.css [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+.xHelpButton.gene-info{
+    padding-top: 5px;
+    margin-right: -16px;
+    padding-left: 10px;
+}
+.suggested.suggest-gene.gene-name{
+    width : 85%;
+}
+.gene-table td, .gene-table th {
+  vertical-align: top;
```
Comment (meatcar) [inline components/widgets/resources/src/main/resources/resources/uicomponents/widgets/extraGeneVariantData.css]: inconsistent indents in this file.

File: components/widgets/resources/src/main/resources/resources/uicomponents/widgets/extraGeneVariantData.js [hunk 10]
```diff
@@ -0,0 +1,524 @@
+var ExtraGeneVariantData = (function (ExtraGeneVariantData) {
+  var tools = ExtraGeneVariantData.tools = ExtraGeneVariantData.tools || {};
+  tools.Editor = Class.create({
+
+    initialize : function () {
+      this.geneClassName = $$('.gene-table.extradata-list')[0].id.substring($$('.gene-table.extradata-list')[0].id.lastIndexOf('-')+1);
+      this.geneVariantClassName = 'PhenoTips.GeneVariantClass';
+
+      //if edit mode
+      if (!!$('inline')) {
```
Comment (meatcar) [inline components/widgets/resources/src/main/resources/resources/uicomponents/widgets/extraGeneVariantData.js]: unnescessary `!!`.

File: components/widgets/resources/src/main/resources/resources/uicomponents/widgets/extraGeneVariantData.js [hunk 11]
```diff
@@ -0,0 +1,524 @@
...
+        var indx = parseInt(item.className.substring(item.className.lastIndexOf('-')+1), 10);
+        varIndex = Math.max(indx, varIndex);
+      });
+      varIndex++;
+
+
+      var data = {GENE_SYMBOL_PLACEHOLDER: geneSymbol, VARIANT_COUNT_PLACEHOLDER: varCount, GENE_INDEX_PLACEHOLDER: geneIndex, VARIANT_INDEX_PLACEHOLDER: varIndex}
+
+      //If we insert first row of variants - create headers
+      if (variantFooter.previous().className == '') {
```
Comment (meatcar) [inline components/widgets/resources/src/main/resources/resources/uicomponents/widgets/extraGeneVariantData.js]: you should use `===` when comparing strings, since `''` can be interpreted as `false` with `==`.

File: components/widgets/resources/src/main/resources/resources/uicomponents/widgets/extraGeneVariantData.js [hunk 12]
```diff
@@ -0,0 +1,524 @@
...
+        var variantIndex = row.className.substring(row.className.lastIndexOf('-')+1);
+        var editVariantLink = new Element('a', {'class' : 'action-edit variant-edit-done v-collapsed', 'href' : '#', id : 'PhenoTips.GeneVariantClass_'+variantIndex+'_editDone'});
+        row.insert(editVariantLink);
+      });
+    },
+
+    createEditButtons : function (className) {
+      $$(className).each(function(row) {
+        var variantIndex = row.className.substring(row.className.lastIndexOf('-')+1);
+        var editVariantLink = new Element('a', {'class' : 'action-edit button secondary variant-edit fa fa-pencil', 'href' : '#', 'title' : "$services.localization.render('PhenoTips.GeneVariantClass.edit.hint')", id : 'PhenoTips.GeneVariantClass_'+variantIndex+'_edit', 'style' : 'font-size: 90%'}).update(' Edit');
```
Comment (meatcar) [inline components/widgets/resources/src/main/resources/resources/uicomponents/widgets/extraGeneVariantData.js]: we enforce 80 char limit on lines of code in java. Maybe we should do the same for javascript? Either way, I think this line is waay too long, and should be broken up for readability. Some other lines in this files are candidates as well.

Comment (antonkatz) [inline components/widgets/resources/src/main/resources/resources/uicomponents/widgets/extraGeneVariantData.js]: +1. Also spaces around and inside ``` + variantIndex + ```

Comment (buske) [inline components/widgets/resources/src/main/resources/resources/uicomponents/widgets/extraGeneVariantData.js]: Wow, there were a lot of jshint errors and warnings. Going to commit a new version shortly.

**Comment by buske** [comment]
Thanks for your comments, @meatcar and @antonkatz. I addressed these in several additional commits. Please take another look and +1 if everything looks okay now.

**Comment by buske** [comment]
This branch was merged via https://github.com/phenotips/phenotips/commit/f36ca6e15fc479ece94845f7370bcdcc16347845. @sdumitriu, shall we close this PR?

**Comment by sdumitriu** [comment]
No, there are a few outstanding commits that need to be merged as well.

**Comment by allasm** [comment]
Outdated and should be closed?

**Comment by veronikaslc** [comment]
Not yet, still has some useful code left there.
